File: 116.PopulatingNextRightPointersinEachNode/clywin123/connect.py
class Solution:
    def connect(self, root):
        if (not root):
            return
        visit = [root]
        while(visit):
            for i in range(len(visit)-1):
                visit[i].next = visit[i+1]
            newvisit = []
            for node in visit:
                if(node.left!=None):
                    newvisit.append(node.left)
                if(node.right!=None):
                    newvisit.append(node.right)
            visit = newvisit[:]

        return

File: 117.PopulatingNextRightPointersinEachNode-II/wyrj/connect.js
/**
 * Definition for binary tree with next pointer.
 * function TreeLinkNode(val) {
 *     this.val = val;
 *     this.left = this.right = this.next = null;
 * }
 */

/**
 * @param {TreeLinkNode} root
 * @return {void} Do not return anything, modify tree in-place instead.
 */
var connect = function(root) {
    if (!root) return;
    root.next = null;
    let parent = root;
    let first = root.left || root.right
    while (first) {
        let curr = first;
        let next = null;
        let nextFirst = curr.left || curr.right;
        while (parent) {
            next = curr === parent.left ? parent.right : (parent.left || parent.right);
            if (!next) parent = parent.next;
            else {
                curr.next = next;
                curr = next;
                if (next === parent.right) parent = parent.next;
                if (!nextFirst) nextFirst = curr.left || curr.right;
            }
        }
        curr.next = null;
        parent = first;
        first = nextFirst;
    }
};
